File: 28.Implement-strStr().go
// https://leetcode.com/problems/remove-duplicates-from-sorted-array/
//
// algorithms
// Easy (32.04%)
// Total Accepted:    430,978
// Total Submissions: 1,345,004
// beats 100.0% of golang submissions

package leetcode

func strStr(haystack string, needle string) int {
	if needle == "" || haystack == needle {
		return 0
	}

	return kmp(haystack, needle)
}

func kmp(source, target string) int {

	sourceLen, targetLen := len(source), len(target)

	if targetLen > sourceLen {
		return -1
	}

	next := genNext(target)
	i, j := 0, 0

	for i < sourceLen && j < targetLen {
		if j == -1 || source[i] == target[j] {
			i++
			j++
		} else {
			j = next[j]
		}
	}

	if j == targetLen {
		return i - targetLen
	}

	return -1
}

func genNext(target string) []int {
	length := len(target)
	next := make([]int, length)
	next[0] = -1

	i, k := 0, -1
	for i < length-1 {
		if k == -1 || target[i] == target[k] {
			k++
			i++
			next[i] = k
		} else {
			k = next[k]
		}
	}

	return next
}

File: 28.Implement-strStr().py
# https://leetcode.com/problems/implement-strstr/
#
# algorithms
# Easy (32.84%)
# Total Accepted:    486,985
# Total Submissions: 1,482,860


class Solution(object):
    def strStr(self, haystack, needle):
        """
        :type haystack: str
        :type needle: str
        :rtype: int
        """
        if needle == '':
            return 0

        len_h, len_n = len(haystack), len(needle)
        if len_n > len_h:
            return -1

        idx_h, idx_n = 0, 0
        next = self.gen_next(needle)

        while idx_h < len_h and idx_n < len_n:
            if idx_n == -1 or needle[idx_n] == haystack[idx_h]:
                idx_n += 1
                idx_h += 1

            else:
                idx_n = next[idx_n]

        if idx_n < len_n:
            return -1

        return idx_h - len_n

    def gen_next(self, s):
        length = len(s)
        next = [0] * length
        next[0] = -1
        j, k = 0, -1

        while j < length - 1:
            if k == -1 or s[j] == s[k]:
                j += 1
                k += 1
                next[j] = k

            else:
                k = next[k]

        return next
